### ISSUE TITLE
Fix implementations for chimney-macro-commons for edge cases in Scala 3

### ISSUE THREAD

**Issue by MateuszKubuszok** [scalalandio/chimney#324]
Scala 2 provides us with some utilities like:

* `tpe.finalResultType`
* `method.typeSignatureIn(tpe)`
* `tpe.substituteTypes(typeParams, typeArgs)`

In Scala 3 we don't have them and the first implementation of cross-platform utilities used by macros don't handle edge cases well:

```scala
case class Foo[A](a: A, rec: Option[Foo[A]])
// when we apply A =:= String we should have properties:
val a: String
val rec: Option[Foo[String]]
// but instead we get:
val a: String
val rec: Option[Foo[A]] // <- A =:= String hasn't been applied
```

```scala
sealed trait Foo[A, B]
case class Impl[C, D](c: C, d: D) extends Foo[D, C]
// when we apply A =:= String, B =:= Int
Foo[String , Int]
// should have a Impl subtype
Impl[Int, String]
```

Another issue is that Scala 2 macros provide is with:

```scala
methodSymbol.isPublic
```

which lets us check if some method or variable would be publicly available. Scala 3 encourages us to check flags:

```scala
!(sym.flags.is(Flags.Private) || sym.flags.is(Flags.PrivateLocal) || sym.flags.is(Flags.Protected))
```

but it doesn't help in situation like:

```scala
class Foo(val a: Int, b: String)
```

where list of symbols containing both `a` and `b` would NOT have any symbol with a private flag, which could lead to assumption that `b` is available, yet an attempt to use `b` would result in compilation error.

This needs to be fixed to unblock Scala 3 release. Code which would have to be modified lives in `chimney-macro-commons`, in `scala-3` folder.

**Comment by MateuszKubuszok** [comment]
It seems that https://github.com/scalalandio/chimney/pull/336 fixed all of the issues we saw during compilation. @jchyb found some issues with symbol ownership with `-Xcheck-macros` and errors format seem to changed (at least in the context of munit, but it might be bug) - these need to be addressed eventually, but they aren't a blocker for preview/milestone release.

**Comment by MateuszKubuszok** [comment]
Released in 0.8.0-RC1.
